src/input/espresso/client.rs

```diff
@@ -0,0 +1,635 @@
+        },
+        testing::TestConfigBuilder,
+    };
+    use staking_cli::{
+        DEV_MNEMONIC,
+        demo::{DelegationConfig, StakingTransactions},
+    };
+    use surf_disco::{Error, StatusCode};
+    use tokio::task::spawn;
+
+    type V = SequencerVersions<MaxSupportedVersion, MaxSupportedVersion>;
+
+    #[test_log::test(tokio::test(flavor = "multi_thread"))]
+    async fn test_epochs() {
+        let port = pick_unused_port().expect("No ports free");
+        let (mut network, _storage) = start_pos_network(port).await;
+
+        let opt = QueryServiceOptions::new(format!("http://localhost:{port}").parse().unwrap());
+        let client = QueryServiceClient::new(opt).await.unwrap();
+
+        // Check that the client's view of the current epoch remains consistent, and wait for an
+        // epoch change.
+        let first_epoch = *network
+            .server
+            .decided_leaf()
+            .await
+            .epoch(EPOCH_HEIGHT)
+            .unwrap();
+        tracing::info!(first_epoch);
+        let next_epoch = loop {
+            // What epoch does the client think it is?
+            let mut epoch_from_client = client.current_epoch().await.unwrap();
+
+            // Grab the expected epoch directly from consensus.
+            let leaf = network.server.decided_leaf().await;
+            let epoch = *leaf.epoch(EPOCH_HEIGHT).unwrap();
+
+            if epoch_from_client != epoch {
+                // It's possible for the consensus epoch to run slightly ahead of the client, both
+                // because we sampled it later, and because the client is ultimately reading from
+                // the node's database, and there is a very small delay in data propagating from the
+                // consensus object to the database. If the client is behind, though, it should not
+                // be far behind, and it should catch up very quickly.
+                assert_eq!(epoch_from_client + 1, epoch);
+                // This should only happen if we have just now changed epochs.
+                assert_eq!(
+                    leaf.height() % EPOCH_HEIGHT,
+                    1,
+                    "client is behind consensus, but we didn't just enter a new epoch at height {}",
+                    leaf.height()
+                );
+                tracing::info!(
+                    epoch_from_client,
+                    epoch,
+                    "client is 1 epoch behind consensus, waiting for it to catch up"
+                );
+                sleep(Duration::from_secs(2)).await;
+                epoch_from_client = client.current_epoch().await.unwrap();
+            }
+
+            tracing::info!(
+                epoch,
+                epoch_from_client,
+                "comparing epoch from client vs epoch from last decided leaf {}",
+                leaf.height()
+            );
+            assert_eq!(epoch_from_client, epoch);
+
+            if epoch > first_epoch {
```

Comment:
I think this check should if epoch > first_epoch + 1, because the first two epochs use default stake tables